compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -337,6 +338,28 @@ impl Context {
 
         inserted_check
     }
+
+    /// True if this is a `Instruction::IncrementRc` or `Instruction::DecrementRc`
+    /// operating on an array directly from a `Instruction::MakeArray` or an
+    /// intrinsic known to return a fresh array.
+    fn is_inc_dec_instruction_on_known_array(
+        instruction: &Instruction,
+        dfg: &DataFlowGraph,
+    ) -> bool {
+        use Instruction::*;
+        if let IncrementRc { value } | DecrementRc { value } = instruction {
+            if let Value::Instruction { instruction, .. } = &dfg[*value] {
+                return match &dfg[*instruction] {
+                    MakeArray { .. } => true,
+                    Call { func, .. } => {
+                        matches!(&dfg[*func], Value::Intrinsic(_) | Value::ForeignFunction(_))
```

Comment:
If it's an array returned by a normal call it wouldn't be a known array, it could be an existing one. This check only returns true for intrinsics which should IIRC create fresh arrays (although perhaps this assumption isn't true? E.g. `slice_push_back`?)

Reply:
> If it's an array returned by a normal call it wouldn't be a known array, it could be an existing one.

Ah yes.

Reply:
Going to merge this for now so I can continue investigation into more rc changes.
I'll open an issue for investigating if this check is correct or not